### PR TITLE
Speaker sample various updates

### PR DESCRIPTION
**Pull request by kul3r4**
- Uses Media Toolkit, Exoplayer and opens a separate screen for Player
- Adds navigation to Player and Volume Screen
- Uses Wear Compose progress ring
- Removes animation

Still need removing Constraint Layout, addressing in future PR.

New version
[SpeakerSamplenew.webm](https://github.com/android/wear-os-samples/assets/820891/a4850a96-a49c-42c8-8dd8-70f947760787)

Old version
[speakersampleapp.webm](https://github.com/android/wear-os-samples/assets/820891/0da1cb48-3e42-430d-a8f5-8729fab8dbed)

### CODE REVIEW

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/ControlDashboard.kt [hunk 1]
```diff
@@ -74,6 +71,10 @@ fun ControlDashboard(
         Spacer(
             modifier = modifier.layoutId(circle)
         )
+        CircularProgressIndicator(
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/ControlDashboard.kt:74]: Nice

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainActivity.kt [hunk 2]
```diff
@@ -18,15 +18,23 @@ package com.example.android.wearable.speaker
...
+import androidx.media3.exoplayer.ExoPlayer
 
 /**
  * A simple activity that just serves as an entry point to [SpeakerApp].
  */
 class MainActivity : ComponentActivity() {
+    @OptIn(UnstableApi::class)
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        val player = ExoPlayer.Builder(this)
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainActivity.kt:32]: This is nice. I suspect we want this in a view model scope?

But I'm not sure.

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SoundRecorder.kt [hunk 3]
```diff
@@ -37,56 +35,7 @@ class SoundRecorder(
...
-            return
-        }
-
-        // Check if there isn't a recording to play
-        if (!audioFile.exists()) return
-
-        suspendCancellableCoroutine { cont ->
-            state = State.PLAYING
-
-            val mediaPlayer = MediaPlayer().apply {
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SoundRecorder.kt:62]: Nice cleanup

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt [hunk 4]
```diff
@@ -46,14 +54,18 @@ import kotlinx.coroutines.launch
  */
 @OptIn(ExperimentalHorologistApi::class)
 @Composable
-fun SpeakerApp() {
+fun SpeakerApp(player: Player) {
     MaterialTheme {
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt:58]: Do we have AppScaffold somewhere?

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    onVolumeClick: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val context = LocalContext.current
+    val activity = context.findActivity()
+
+    val volumeUiState by volumeViewModel.volumeUiState.collectAsStateWithLifecycle()
+    val playerViewModel = SpeakerPlayerViewModel(PlayerRepositoryImpl(), player, activity)
+
+    PlayerScreen(
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt:52]: Nice

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.android.horologist.media.model.Media
+import com.google.android.horologist.media.ui.state.PlayerViewModel
+import java.io.File
+import kotlinx.coroutines.launch
+
+@OptIn(ExperimentalHorologistApi::class)
+class SpeakerPlayerViewModel(
+    playerRepository: PlayerRepositoryImpl,
+    player: Player,
+    activity: Context
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerViewModel.kt:32]: This is probably a leak.

**Review by yschimke — COMMENTED** [review]
Looks great, but mixing activity and view model is a code smell.

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt (hunk 4 above)
Comment (kul3r4) [reply WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt:58]: yeah,  [Speaker app](https://github.com/android/wear-os-samples/pull/1094/files#diff-667f2835722c1429eaeb80be39a23315fd175d0c299403e0bd99ec7c73825d1dR86)

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerViewModel.kt (hunk 6 above)
Comment (kul3r4) [reply WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerViewModel.kt]: PTAL, I am passing the File and the URI now.

**Comment by kul3r4** [comment]
> Looks great, but mixing activity and view model is a code smell.

Thanks! PTAL I removed the context from the viewModel

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+) {
+    val context = LocalContext.current
+    val activity = context.findActivity()
+
+    val volumeUiState by volumeViewModel.volumeUiState.collectAsStateWithLifecycle()
+
+    val outputFileName = "audiorecord.opus"
+    val audioFile = File(activity.filesDir, outputFileName)
+    val audioFileUri = activity.filesDir.path + "/" + outputFileName
+    val playerViewModel = SpeakerPlayerViewModel(
```
Comment (yschimke) [inline WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt:55]: You can't create view models like this.

If you just need it from the VM, the view model can use the application context.

Comment (kul3r4) [reply WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt]: What about using a factory? https://github.com/android/wear-os-samples/pull/1094/files#diff-02cbeefda5435ff1dca10c8c0961531257725dfe73a1caa6227fafb62ea7b40bR61

Comment (yschimke) [reply WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt]: Yep, but what about just passing in filename? Do the filesDir lookup in the view model?

Comment (yschimke) [reply WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerPlayerScreen.kt]: Maybe not needed if it's hardcoded

File: WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainActivity.kt (hunk 2 above)
Comment (kul3r4) [reply WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainActivity.kt]: it's now in the view model

**Review by yschimke — APPROVED** [review]

**Pull request merged**